cloud/linode/instances.go

```diff
@@ -29,176 +29,89 @@ func (e instanceNoIPAddressesError) Error() string {
 	return fmt.Sprintf("instance %d has no IP addresses", e.id)
 }
 
-func (i *instances) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.NodeAddress, error) {
-	ctx = sentry.SetHubOnContext(ctx)
-	sentry.SetTag(ctx, "node_name", string(name))
-
-	linode, err := linodeByName(ctx, i.client, name)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return nil, err
-	}
-
-	addresses, err := i.nodeAddresses(ctx, linode)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return nil, err
-	}
-
-	return addresses, nil
-}
+func (i *instances) lookupLinode(ctx context.Context, node *v1.Node) (*linodego.Instance, error) {
+	providerID := node.Spec.ProviderID
+	nodeName := types.NodeName(node.Name)
 
-func (i *instances) NodeAddressesByProviderID(ctx context.Context, providerID string) ([]v1.NodeAddress, error) {
-	ctx = sentry.SetHubOnContext(ctx)
 	sentry.SetTag(ctx, "provider_id", providerID)
+	sentry.SetTag(ctx, "node_name", node.Name)
 
-	id, err := parseProviderID(providerID)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return nil, err
-	}
-
-	linode, err := linodeByID(ctx, i.client, id)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return nil, err
-	}
-
-	addresses, err := i.nodeAddresses(ctx, linode)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return nil, err
-	}
-
-	return addresses, nil
-}
-
-func (i *instances) nodeAddresses(ctx context.Context, linode *linodego.Instance) ([]v1.NodeAddress, error) {
-	var addresses []v1.NodeAddress
-	addresses = append(addresses, v1.NodeAddress{Type: v1.NodeHostName, Address: linode.Label})
-
-	ips, err := i.client.GetInstanceIPAddresses(ctx, linode.ID)
-	if err != nil {
-		return nil, err
-	}
-
-	if (len(ips.IPv4.Public) == 0) && (len(ips.IPv4.Private) == 0) {
-		return nil, instanceNoIPAddressesError{linode.ID}
-	}
-
-	if len(ips.IPv4.Public) > 0 {
-		addresses = append(addresses, v1.NodeAddress{Type: v1.NodeExternalIP, Address: ips.IPv4.Public[0].Address})
-	}
-	// Allow Nodes to not have an ExternalIP, if this proves problematic this will be reverted
+	if providerID != "" {
+		id, err := parseProviderID(providerID)
+		if err != nil {
+			sentry.CaptureError(ctx, err)
+			return nil, err
+		}
+		sentry.SetTag(ctx, "linode_id", strconv.Itoa(id))
 
-	if len(ips.IPv4.Private) > 0 {
-		addresses = append(addresses, v1.NodeAddress{Type: v1.NodeInternalIP, Address: ips.IPv4.Private[0].Address})
+		return linodeByID(ctx, i.client, id)
 	}
-	// Allow Nodes to not have an InternalIP, if this proves problematic this will be reverted
 
-	return addresses, nil
+	return linodeByName(ctx, i.client, nodeName)
 }
 
-func (i *instances) InstanceID(ctx context.Context, nodeName types.NodeName) (string, error) {
+func (i *instances) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
```

Comment:
Can we rename this function to be in the form of a true / false question since this is returning a bool.

I like things that start with is, has, can, was.

Something like
```suggestion
func (i *instances) HasInstance(ctx context.Context, node *v1.Node) (bool, error) {
```

Reply:
We cannot rename this method, because then we wouldn't implement the [upstream interface](https://pkg.go.dev/k8s.io/cloud-provider#InstancesV2).